xdsl/dialects/x86/register.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+from abc import ABC
+from collections.abc import Sequence
+
+from xdsl.backend.register_type import RegisterType
+from xdsl.ir import Attribute
+from xdsl.irdl import (
+    irdl_attr_definition,
+)
+from xdsl.parser import AttrParser
+from xdsl.utils.exceptions import VerifyException
+
+
+class X86RegisterType(RegisterType):
+    """
+    An x86 register type.
+    """
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[Attribute]:
+        with parser.in_angle_brackets():
+            name = parser.parse_optional_identifier()
+            if name is not None:
+                if not name.startswith("e") and not name.startswith("r"):
+                    assert name in cls.abi_index_by_name(), f"{name}"
+            else:
+                name = ""
+            return cls._parameters_from_spelling(name)
+
+    def verify(self) -> None:
+        name = self.spelling.data
+        if not self.is_allocated or name.startswith("e") or name.startswith("r"):
+            return
+        if name not in type(self).abi_index_by_name():
+            raise VerifyException(f"{name} not in {self.instruction_set_name()}")
+
+
+X86_INDEX_BY_NAME = {
+    "rax": 0,
+    "rcx": 1,
+    "rdx": 2,
+    "rbx": 3,
+    "rsp": 4,
+    "rbp": 5,
+    "rsi": 6,
+    "rdi": 7,
+    "r8": 8,
+    "r9": 9,
+    "r10": 10,
+    "r11": 11,
+    "r12": 12,
+    "r13": 13,
+    "r14": 14,
+    "r15": 15,
+}
+
+
+@irdl_attr_definition
+class GeneralRegisterType(X86RegisterType):
+    """
+    An x86 register type.
+    """
+
+    name = "x86.reg"
+
+    @classmethod
+    def unallocated(cls) -> GeneralRegisterType:
+        return Registers.UNALLOCATED_GENERAL
+
+    @classmethod
+    def instruction_set_name(cls) -> str:
+        return "x86"
+
+    @classmethod
+    def abi_index_by_name(cls) -> dict[str, int]:
+        return X86_INDEX_BY_NAME
+
+
+class Registers(ABC):
```

Comment:
I'd recommend getting rid of this and placing the constants at the top level of this file

Reply:
what are you referring to by "this"? Do you mean getting rid of the class structure and placing the registers as loose constants?